### PR TITLE
[Merged by Bors] - feat({ Tactic + test }/MoveAdd): make `And/Or` left-associative

### PR DESCRIPTION
**Pull request by adomani**
In this [Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/move_oper.20.20HAnd.2EhAnd.3F/near/403820462) it was pointed out that `move_oper` always assumes right-associativity also for `And/Or`.

This PR makes `move_oper` left-associative on `And/Or`.

Affected files:
```bash
Tactic/MoveAdd
test/MoveAdd
```

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/MoveAdd.lean [hunk 1]
```diff
@@ -223,10 +223,11 @@ If `exs` is the list `[e₁, e₂, ..., eₙ]` of `Expr`essions, then `sumList p
 `prepOp (prepOp( ... prepOp (prepOp e₁ e₂) e₃) ... eₙ)`.
 -/
 partial
-def sumList (prepOp : Expr) : List Expr → Expr
```
Comment (eric-wieser) [inline Mathlib/Tactic/MoveAdd.lean:226]: Can you document this new boolean argument?

Comment (eric-wieser) [reply Mathlib/Tactic/MoveAdd.lean:226]: (i'd be Inclined to move it outside the pattern match so that you can name it.

Comment (adomani) [reply Mathlib/Tactic/MoveAdd.lean:226]: Documentation added and also I pulled out the new `Bool`ean argument out of the pattern match.

File: Mathlib/Tactic/MoveAdd.lean [hunk 2]
```diff
@@ -215,18 +215,25 @@ def prepareOp (sum : Expr) : Expr :=
...
+def sumList (prepOp : Expr) (left_assoc? : Bool) : List Expr → Expr
   | []    => default
-  | [a]    => a
-  | a::as => as.foldl (fun x y => Expr.app (prepOp.app x) y) a
+  | [a]   => a
+  | a::as =>
+    if left_assoc? then
+      Expr.app (prepOp.app a) (sumList prepOp true as)
```
Comment (eric-wieser) [inline Mathlib/Tactic/MoveAdd.lean:234]: Or alternatively (untested):
```suggestion
      as.foldr (fun x y => Expr.app x (prepOp.app y)) a
```

Comment (adomani) [reply Mathlib/Tactic/MoveAdd.lean:234]: I tried this, but I think that it does not work: the `a` should really be the *last* element of the list.  At least, this is what my understanding of `foldr` does.  So, there may be some more idiomatic way of doing this, but the following would be closer to what it should be:
```lean
  | l@(a::as) =>
    if left_assoc? then
      (l.dropLast).foldr (fun x y => Expr.app (prepOp.app x) y) l.getLast!
```

**Comment by robertylewis** [comment]
Thanks, and sorry for the delay!

bors merge

**Pull request closed without merging**